dataproc/dataproc.go

```diff
@@ -0,0 +1,304 @@
+		log.Fatal(err)
+	}
+
+	// Wait for the job to complete
+	log.Println("Waiting for job to complete")
+	if _, err = waitForJob(service, jobID, &cluster); err != nil {
+		log.Fatal(err)
+	}
+	log.Println("Job is finished")
+
+	output, err := getJobOutput(storageService, jobID, &cluster)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("Job output:\n%s\n", output)
+
+	// Delete the cluster
+	if _, err = deleteCluster(service, &cluster); err != nil {
+		log.Fatal(err)
+	}
+	log.Println("Cluster deleted")
+}
+
+// createCluster creates a Cloud Dataproc cluster with the given name and region.
+func createCluster(service *dataproc.Service, cluster *clusterConfig) (response string, err error) {
+	// Create a gceConfig object for the cluster
+	gceConfig := dataproc.GceClusterConfig{
+		ZoneUri: fmt.Sprintf(computeUriBase, cluster.project, cluster.zone),
+	}
+
+	// Create a (Dataproc API) clusterConfig for the cluster
+	clusterConfig := dataproc.ClusterConfig{
+		GceClusterConfig: &gceConfig,
+	}
+
+	// Create a cluster object
+	clusterSpec := dataproc.Cluster{
+		ClusterName: cluster.name,
+		ProjectId:   cluster.project,
+		Config:      &clusterConfig,
+	}
+
+	// Create the cluster
+	res, err := service.Projects.Regions.Clusters.Create(cluster.project, cluster.region, &clusterSpec).Do()
+
+	return fmt.Sprintf("%s", res), err
+}
+
+// deleteCluster deletes the Cloud Dataproc cluster with the given project, region, and name.
+func deleteCluster(service *dataproc.Service, cluster *clusterConfig) (response string, err error) {
+	// Delete the cluster
+	res, err := service.Projects.Regions.Clusters.Delete(cluster.project, cluster.region, cluster.name).Do()
+
+	return fmt.Sprintf("%s", res), err
+}
+
+// getJobOutput returns the text from the job (raw driver output) with the given project, cluser name, bucket id, and job id.
+func getJobOutput(storageClient *storage.Client, job string, cluster *clusterConfig) (output []byte, err error) {
+
+	// Format the object name based on the Cloud Dataproc service's GCS logging
+	// see https://cloud.google.com/dataproc/concepts/driver-output for details
+	object := fmt.Sprintf("google-cloud-dataproc-metainfo/%s/jobs/%s/driveroutput.000000000", cluster.uuid, job)
+
+	// Read the file
+	rc, err := storageClient.Bucket(cluster.bucket).Object(object).NewReader(context.Background())
+	output, err = ioutil.ReadAll(rc)
+	rc.Close()
+
+	return output, err
+}
+
+// getJobStatus returns the status of the job with the given job id, project, and region.
+func getJobStatus(service *dataproc.Service, jobID string, project string, region string) (status string, err error) {
+	// Get the Job's status
+	res, err := service.Projects.Regions.Jobs.Get(project, region, jobID).Do()
+	if err != nil {
+		return "", err
+	}
+	return res.Status.State, nil
+}
+
+// listClusters lists all clusters in the current project.
+func listClusters(service *dataproc.Service, project string, region string) (clusters [][]string, err error) {
```

Comment:
Refactored to introduce a second struct `clusterDetails` to contain details about a created cluster. This cleanly (I think) separates a `config` which is used to create a cluster with the actual returned details from a cluster.

Unfortuinately, our API does not cleanly return `region` or `zone` so I have to parse some URLs to get them.